tests/components/transmission/conftest.py

```diff
@@ -0,0 +1,138 @@
+"""Fixtures for the Transmission integration tests."""
+import json
+
+import pytest
+from transmissionrpc.error import TransmissionError
+
+from homeassistant.components.transmission import (
+    CONF_HOST,
+    CONF_LIMIT,
+    CONF_NAME,
+    CONF_ORDER,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_USERNAME,
+    DOMAIN,
+)
+from homeassistant.components.transmission.const import ORDER_OLDEST_FIRST
+
+from tests.async_mock import patch
+from tests.common import MockConfigEntry, load_fixture
+
+MOCK_TORRENTS_OLDEST_TO_NEWEST = [2, 1, 3, 4, 5]
+MOCK_TORRENTS_BEST_TO_WORST = [3, 1, 5, 2, 4]
+MOCK_LIMIT_INITIAL = 20
+MOCK_LIMIT_TRUNCATED = 2
+
+
+def setup_transmission(
+    hass,
+    name="Transmission",
+    host="0.0.0.0",
+    username="user",
+    password="pass",
+    port=9091,
+    limit=MOCK_LIMIT_INITIAL,
+    order=ORDER_OLDEST_FIRST,
+):
+    """Init the mock Transmission entry."""
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={
+            CONF_NAME: name,
+            CONF_HOST: host,
+            CONF_USERNAME: username,
+            CONF_PASSWORD: password,
+            CONF_PORT: port,
+            CONF_LIMIT: limit,
+            CONF_ORDER: order,
+        },
+    )
+    entry.add_to_hass(hass)
+    return entry
+
+
+def get_correct_fixture(*args):
+    """Return the appropriate JSON string."""
+    method = json.loads(args[0])["method"]
+    if method == "torrent-get":
+        return load_fixture("transmission-torrents-get.json")
+    if method == "session-get":
+        return load_fixture("transmission-session-get.json")
+    if method == "session-stats":
+        return load_fixture("transmission-session-stats.json")
+    return ""
+
+
+async def mock_client_setup(hass, entry=None, **kwargs):
+    """Mock Transmission client setup."""
+    # pylint: disable=protected-access
+    if entry is None:
+        entry = setup_transmission(hass)
+    with patch(
+        "homeassistant.components.transmission.transmissionrpc.Client._http_query",
+        side_effect=get_correct_fixture,
+    ):
+        await hass.config_entries.async_setup(entry.entry_id)
+        for name, value in kwargs.items():
+            setattr(hass.data[DOMAIN][entry.entry_id]._tm_data.data, name, value)
```

Comment:
Please don't interact with integration details in the tests. We just want to use core interfaces to set up the tests and patches on the library to mock different behaviors.

Reply:
I am not certain how to go about this.  The setattr is overwriting an attribute of the dependency instance that is part of the integration instance.

Here is a pattern that makes this a bit more explicit - convert the mock_client_setup to a fixture that is yielded from and back to:
```python3
@pytest.fixture(name="tm_client")
async def mock_client_setup(hass):
    """Mock Transmission client setup."""
    # pylint: disable=protected-access
    if entry is None:
        entry = setup_transmission(hass)
    with patch(
        "homeassistant.components.transmission.transmissionrpc.Client._http_query",
        side_effect=get_correct_fixture,
    ):
        await hass.config_entries.async_setup(entry.entry_id)
        yield hass.data[DOMAIN][entry.entry_id]._tm_data.data  # this is not OK?
        await hass.async_block_till_done()
```
And then in test_sensor.py we still explicitly set the attribute so we know what is being tested.  It is the case that we are testing a function that interprets the value of download speed to come up with "Seeding".
```python3
async def test_status_seeding(hass, tm_client):
    """Test the Transmission Status Seeding."""
    setattr(tm_client, "downloadSpeed", 0)
    yield  # though I could just as well include the await hass.async_block_till_done() statement here

    status = hass.states.get("sensor.transmission_status")
    assert status.state == "Seeding"
   ```

The obvious way to avoid my hack is to to have a separate JSON fixture for every change but this would be cumbersome.  I first experimented with "intercepting the JSON" but it did not seem good, lol!

Reply:
OK, I thought more.  It looks like I can patch the library where it would be storing the data it has retrieved from the JSON.